7368 missing sql embed backwards compatibility

javitonino opened the pull request:
Fixes #7368 

Instead of trying to add the `query` when using the old embed with analyses, this modified the old embed to use named maps when showing a visualization with analyses.

javitonino commented:
CR @juanignaciosl. This is a different approach: instead of returning the query in the vizjson2, it uses named_maps when using analyses (given that all maps with analyses have named maps, which I think is right). Does this makes sense? If not, I can revert to the original approach.

juanignaciosl commented:
That's the approach for the new editor. It might make sense and be far more elegant. What do you think, @javisantana ?

javisantana commented:
if you are developing an application with cartodb.js using the viz.json, you go to the new editor, create an analysis, will the viz.json v2 will be sent with the SQL?

javitonino commented:
@javisantana No, with this change, if there is an analysis, it always return a namedmap layer, which has no SQL in it.

I have an alternative patch that includes the SQL, but it has the problem that has different behaviour for different table privacies (expected):
- Public dataset (no named maps): Base data layer is shown (no analysis), includes SQL
- Private datasets (named maps): Analysis is shown, no SQL query.

javisantana commented:
Ok, your change seems good enough to have retrocompatibility for some time.

what do you think @juanignaciosl

juanignaciosl commented on app/models/visualization/member.rb:
```diff
@@ -493,7 +493,7 @@ def has_private_tables?
       # Despite storing always a named map, no need to retrieve it for "public" visualizations
       def retrieve_named_map?
-        password_protected? || has_private_tables?
+        password_protected? || has_private_tables? || Carto::Analysis.where(visualization_id: id).any?
```
Since it won't be cached please check that it isn't triggered too many times for common scenarios (dashboard / embed / public), please.

javitonino replied:
This is only called once from the vizjson presenters. As the vizjson itself is cached, this should be alright.

juanignaciosl commented:
I've made a comment about a minor performance risk on the Sequel model. If this works for the known scenario for embeds and cartodb.js applications, 👍

javitonino commented:
Works for embed and cartodb.js. The only issue I can anticipate is that, as this is returning named maps, the user won't be able to manually specify the SQL in cartodb.js when using a viz with analyses (it returns a meaningful error instead).

javisantana commented:
@javitonino that's an issue because an app could break just because you open the new editor

javitonino commented:
Ok, after taking comments into account, I've rewritten it so it generates the old vizjson (no named maps, includes `sql`). As we still need to include the `source` for named maps but not for public vizjson, I've added an option to the vizjson presenter so it is easier to tell them apart. Please CR @juanignaciosl

juanignaciosl commented on app/models/layer/presenter.rb:
```diff
@@ -203,12 +203,12 @@ def options_data_v2
           }
           source = layer.options['source']
-          if source
+          if options[:for_named_map] && source
```
Welcome to the Dirty Hack Party 💃

juanignaciosl commented:
We both know that it's a dirty hack, but I think that it's currently our best shot. Code is ok, but please try it as much as you can, I know that it's cumbersome but the integration with the Maps API is fragile. 👍

javitonino commented:
After extensive testing: this works for all combinations of table privacy, editor version and embed versions (and cartodb.js) for normal data layers. 
There are multiple problems with Torque layers, but they can be tracked in other issues:
- Adding an analysis to a Torque layers resets the styles without changing the layer type, breaking the visualization (@xavijam is tracking this).
- An old-editor map with a Torque layer and a normal data layer won't load in the new embed. #7422

I say merge this and start looking at Torque in the other issues.
